Fix #460: Add OpenID connect support

magopian opened the pull request:
Fixes #460

magopian commented:
The looks will have to be improved: I wanted to have multiple submit buttons (one for each openID provider), but that's not possible with react-json-schema-form. Well, you can have multiple submit buttons, but you won't know which one was pressed (the `value` isn't passed down after the validation it seems).

For now, this is how it looks:

<img width="569" alt="screen shot 2018-03-30 at 16 40 43" src="https://user-images.githubusercontent.com/167767/38141753-f0b36c6a-3439-11e8-968d-30677828e80f.png">

No "auth0 dance" is implemented yet.

leplatrem commented:
Cool, that's a start!

In our use case we'll have only one provider, so for that particular case we case even hide the radio

leplatrem commented on src/client.js:
```diff
@@ -8,7 +8,8 @@ let client: ?KintoClient;
   switch (auth.authType) {
-    case "fxa": {
+    case "fxa":
+    case "openid": {
```
Ideally the `Bearer` string should be taken from the capability

leplatrem replied:
@magopian can you open a follow-up issue for this please?

magopian replied:
opened #543

magopian commented:
This currently doesn't work, check the TODOs in the code.
Multiple issues:
1/ we overwrite the "server info" from the authentified request we do when being redirected after an auth0 login dance, with the "server info" from the default server that is displayed when we mount the AuthForm component
2/ when we get the user id back from the server, it looks like `<provider>:<token>`. But what we're expecting is `<authType>:<token>`. So we receive eg `auth0:token` instead of `openid:token`, in `setupSession`

leplatrem commented:
I was thinking of something: instead of having two list of bullets points, we could merge the OpenID method and the list of provider into a single list, like this:
- Anonymous
- Basic Auth
- Google (OpenID)
- Auth0 (OpenID)
- FxA (OpenID)

magopian commented:
So, the current state of this PR is: it should be working, from what I've tested locally with a local kinto server using the configuration from https://github.com/leplatrem/kinto-oidc-demo#configure--run-kinto, and the google openid provider.

There's a known caveat though: when logging out, the server capabilities aren't fetched again (and are removed from the session), so the supported auth methods aren't displayed until the server is changed or the page is refreshed. This is because it's a nightmare of race conditions and I can't find where and when to get the server information in order to cover all the edge cases.

There's still a bunch of debug messages (`console.log`) littered in the code, and probably some missing tests, but it's in a state where it should be testable. I also haven't addressed your comment https://github.com/Kinto/kinto-admin/pull/509#discussion_r178312847 yet.

What do you think @leplatrem ?

magopian commented:
I added a test, and also implemented your comment about having one auth method per provider "inlined":

<img width="558" alt="screen shot 2018-05-18 at 10 25 12" src="https://user-images.githubusercontent.com/167767/40234608-ca2a53ee-5aa7-11e8-9f6c-45995c42e635.png">

leplatrem commented:
Nice!

I will look at the code in details next week.

@peterbe could you try it with your Kinto instance configured with Auth0?

Regarding the logout bug, it's not a critical failure. We could ship the feature and mark it as *experimental* while we work on it. It would allow us to slowly go forward with the OpenID deployment for remote settings, and more importantly, take over what you started to give you some relief :)

peterbe commented:
Who the heck logs out anyway? The only time I ever log out is my online bank and when I realize the mistake of logging in with peterbe@m.c when I should have logged in with pbengtsson@m-c. :)

peterbe commented:
<img width="1616" alt="screen shot 2018-05-18 at 10 27 44 am" src="https://user-images.githubusercontent.com/26739/40240258-2af48dba-5a86-11e8-9a9e-dcf994c4b8e2.png">

Right after the redirect from Auth0.com's log in screen, there was a page with lots of ugly debugging info near the top of the page. Then it redirected to what it looks like in this screenshot.

leplatrem commented:
> Then it redirected to what it looks like in this screenshot.

:tada:

magopian commented:
I'm not sure I understand your comment @peterbe? Is there something wrong in the screenshot you see?
From what I see, everything seems to have worked correctly. The debug output is probably linked to my comment in https://github.com/Kinto/kinto-admin/pull/509#issuecomment-389901873 : there's still a bunch of `console.log` that's left in the code while I make sure this feature is working as expected.

Once you folks are happy with what's here, I'll clean it up ;)

peterbe commented:
> Is there something wrong in the screenshot you see?

Half a second before this screenshot, right after being redirected back from Auth0 the screen looked very different. It flashed by so quickly I couldn't really see what it said. It looked like a piece of JSON being displayed in the blue flash message near the top. I'll try to make it happen again. 
I'm not referring to `console.log` debug messages. It was in the DOM.

peterbe commented:
@magopian Actually, instead of me trying to reproduce it, and probably failing to make a screenshot, how about I send you my kinto.ini (with my personal Auth0 credentials) to you and you can try it yourself locally?

magopian commented:
I think I can now reproduce this issue: when we get back from the auth0 dance, we land on an url like

http://localhost:3000/#/auth/<token payload>

And the payload is displayed in the breadcrumbs for some reason, and appears like that:

<img width="532" alt="screen shot 2018-05-22 at 15 06 38" src="https://user-images.githubusercontent.com/167767/40364168-eade8a24-5dd1-11e8-914b-846535d0c87d.png">


Is that what you saw @peterbe ?

magopian commented:
@peterbe ok I believe I've fixed the issue: I've excluded the `auth` crumb in the breadcrumbs.

peterbe commented:
Yes! That's what my poor eyes had to suffer through.

magopian commented:
ok great, so I'll just clean the PR, see if I can add a couple more tests maybe, and ask you if you're happy with me merging it 👍

leplatrem commented on README.md:
```diff
@@ -81,7 +81,7 @@ The `KintoAdmin` component accepts a `settings` prop, where you can define the f
 - `maxPerPage`: The max number of results per page in lists (default: `200`).
 - `singleServer`: The server URL to be used (default: `null`). This removes the ability to connect to multiple servers.
-- `authMethods`: The list of supported authenticated methods (default: `["basicauth", "ldap", "fxa"]`). The first is the default one.
+- `authMethods`: The list of supported authenticated methods (default: `["basicauth", "ldap", "fxa", "portier", "openid"]`). The first is the default one.
```
Is `authMethods` still a thing?

leplatrem commented on src/actions/session.js:
```diff
@@ -155,6 +155,23 @@ function postToPortier(server: string, redirect: string): NavigationResult {
+  const redirect = `${origin}${pathname}#/auth/${payload}/`;
+  document.location.href = `${strippedServer}/${strippedAuthPath}?callback=${encodeURIComponent(
+    redirect
+  )}&scope=openid email`;
```
nit: maybe use intermediary variables for callback and scope to avoid wrapping?

peterbe replied:
I think it's Pretty :)

leplatrem commented on src/components/AuthForm.js:
```diff
@@ -178,133 +185,147 @@ const loginPasswordUiSchema = {
+    portier: "Portier",
+  };
+  if (authType.startsWith("openid-")) {
+    const provider = authType.replace("openid-", "");
```
Maybe add a comment here that we construct this `openid-*` string in `getSupportedAuthMethods()`

leplatrem commented on src/components/AuthForm.js:
```diff
@@ -442,15 +468,40 @@ export default class AuthForm extends PureComponent<
+        if (!providerData) {
+          throw "Couldn't find provider data in the state. Bad.";
+        }
+        console.log("providerData", providerData);
```
leftover or intentional info?

leplatrem commented on src/components/AuthForm.js:
```diff
@@ -442,15 +468,40 @@ export default class AuthForm extends PureComponent<
+        } = session;
+        const providerData = providers.find(p => p.name === openidProvider);
+        if (!providerData) {
+          throw "Couldn't find provider data in the state. Bad.";
```
use `throw new Error(str)`

leplatrem commented on src/sagas/session.js:
```diff
@@ -84,9 +92,16 @@ export function* setupSession(
+      (!userId ||
+        (!userId.startsWith(authType + ":") &&
+          // If the authType is openid, the userId doesn't start with "openid:" but "auth0:".
+          (authType.startsWith("openid-") && !userId.startsWith("auth0:"))))
```
I don't understand this test.

The user id starts with the provider name (the right part of authType if I understand correctly)

magopian replied:
are you sure about that? In my tests, i got `auth0:...` back, which was a surprise to me!

magopian replied:
seems my issue with the `auth0:` prefix was due to some configuration problem for the kinto running locally... I've fixed that (and the test)

leplatrem commented on test/sagas/session_test.js:
```diff
@@ -207,6 +233,31 @@ describe("session sagas", () => {
+        const serverInfo = {
+          ...serverInfo,
+          user: { id: "auth0:token" },
+        };
```
I would say this situation is impossible to obtain with a real server.

If you authenticate with a provider name `google`, then your userid will be `google:tartempion`

magopian replied:
are you sure about that? In my tests, i got `auth0:...` back, which was a surprise to me!

peterbe commented on src/actions/session.js:
```diff
@@ -155,6 +155,23 @@ function postToPortier(server: string, redirect: string): NavigationResult {
+  const { origin, pathname } = document.location;
+  const { server } = authFormData;
+  const strippedServer = server.replace(/\/$/, "");
+  const { auth_path } = provider;
```
nit: This one's under_score but the next one is camelCase. I.e. `auth_path` and `[stripped]AuthPath`.

peterbe commented on src/components/AuthForm.js:
```diff
@@ -178,133 +185,147 @@ const loginPasswordUiSchema = {
-  fxa: "Firefox Account",
-  ldap: "LDAP",
-  portier: "Portier",
+const authLabels = authType => {
```
Shouldn't this be called `authLabel` (or `getAuthLabel`)?

peterbe commented on src/components/AuthForm.js:
```diff
@@ -178,133 +185,147 @@ const loginPasswordUiSchema = {
+    const provider = authType.replace("openid-", "");
+    const prettyProvider =
+      labels[provider] || provider.charAt(0).toUpperCase() + provider.slice(1);
+    return `${prettyProvider} (OpenID)`;
```
If `authType` is something like `openid-Auth0.com` (aka `openid-${provider.name}`) then the name *is* `authType.replace("openid-", "");`

I think it should be the other way around. I.e.
```javascript
const providerName = authType.replace("openid-", "");
return `OpenID Connect (${provider})`
```

*Note!* It's called "OpenID Connect", not "OpenID". "OpenID" was some other junk we tried and failed 10-15 years ago. "OpenID Connect" is a OAuth2 sub-standard.

peterbe commented on src/components/AuthForm.js:
```diff
@@ -414,24 +436,28 @@ export default class AuthForm extends PureComponent<
-      authType
-    )
+    const { uiSchema } = authSchemas(authType);
+    let { schema } = authSchemas(authType);
```
Why `let` and not `const`? I don't see it changing.

peterbe commented:
<img width="409" alt="screen shot 2018-05-23 at 8 06 24 am" src="https://user-images.githubusercontent.com/26739/40423344-3c732c30-5e60-11e8-983c-4bc65242ea90.png">

Nice that it caught my typo and handled the Network error gracefully.

peterbe commented:
By the way, I re-tested actually logging in using my local server and it still works. 
I must admit that I haven't really the discussion about the prefix on the email. Like, why am I logged in as `auth0:peterbe@mozilla.com` and not `peterbe@mozilla.com`?

peterbe commented:
When I click "Logout", I get logged out but now the known providers are lost:

<img width="1616" alt="screen shot 2018-05-23 at 8 12 24 am" src="https://user-images.githubusercontent.com/26739/40423613-20f6347e-5e61-11e8-8e76-c2072de2da0b.png">

I have to put my cursor into the Server field, make a change (e.g. editing it to something different and editing back to what it should be) and then the two options appear again.

peterbe commented:
> If that something you'd like to change or optionally disable, we should discuss it there :)

That got more complicated than I had hoped. I guess if kinto-admin is for techy super users it's OK to "butcher" my email address. All I know is that my email is `peterbe@mozilla.com` not `auth0:peterbe@mozilla.com`. But I can see the value in including info about how I logged in. Especially since it plants in my head that the next time I log in, I can remember that the last time I used OpenID Connect and not FxA.

leplatrem commented:
> I guess if kinto-admin is for techy super users it's OK to "butcher" my email address. All I know is that my email is peterbe@mozilla.com not auth0:peterbe@mozilla.com.

`auth0:peterbe@mozilla.com` is not your email address. It's your user id.

And BTW, your user id looks like an email address because you configured your server [in some specific way](https://kinto.readthedocs.io/en/stable/configuration/settings.html#openid-connect).

But let's discuss that elsewhere :)

peterbe commented:
> auth0:peterbe@mozilla.com is not your email address. It's your user id.

I know I know :) I was just playing devil's advocate. It just looks a bit odd when so many other (consumer) websites display my email address in a regular way. Granted, so many other (consumer) websites display it as "Peter Bengtsson" instead :)

magopian commented:
> When I click "Logout", I get logged out but now the known providers are lost:
> 
> ...
> 
> I have to put my cursor into the Server field, make a change (e.g. editing it to something different and editing back to what it should be) and then the two options appear again.

@peterbe this is what I was referring as the caveat in https://github.com/Kinto/kinto-admin/pull/509#issuecomment-389901873

I believe I've fixed the nits and renamed `Google (openID)` to `OpenID Connect (google)`.

magopian commented:
Should I merge this? Is this happening?
